### PR TITLE
Trim trivial qubits from circuit and Hamiltonian

### PR DESCRIPTION
**Pull request by elloyd-1qbit**
Given a circuit and Hamiltonian these functions will:
- Calculate expectation values on qubits which are unentangled and in the |0> or |1> state.  
- Incorporate these expectation values (either 0, +/-1) into the Hamiltonian, potentially reducing the number of terms.

### CODE REVIEW

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            continue
+        # Calculate state of single qubit clifford circuits, ideally this would be done with a clifford simulator
+        # for now only look at first two gate combinations typical of the QMF state in QCC methods
+        num_gates = len(circ._gates)
+        if num_gates not in (1, 2):
+            circuit_new += circ
+            continue
+
+        gate0, gate1 = circ._gates[:2] + [None] * (2 - num_gates)
+        gate_0_is_clifford = True if gate0 is not None and (gate0.parameter == "" or np.isclose(gate0.parameter, np.pi)) else False
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:96]: If this applies to any multiple of pi, I'd suggest upgrading to
`np.isclose(gate0.parameter%np.pi, 0)`

Also, your whole `if` statement is already a True / False expression ! You can just return it directly here:

`gate_0_is_clifford = (gate0 is not None and (gate0.parameter == "" or np.isclose(...)))`

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        gate_1_is_clifford = True if gate1 is not None and (gate1.parameter == "" or np.isclose(gate1.parameter, np.pi)) else False
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:97]: same as above

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    """
+        Splits Circuit into entangled. and unentangled components.
+        Returns entangled Circuit, and the indices and states of unentangled qubits
+
+        Args:
+            circuit (Circuit): circuit to be trimmed
+        Returns:
+            circuit_new (Circuit) : Trimmed, entangled circuit
+            trim_states (list): state of removed qubits, 0 or 1
+            trim_index (list):  index of removed qubits
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:67]: indices

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+def trim_trivial_qubits(operator, circuit):
+    """
+        Trims circuit and operator based on expectation values calculated from
+        trivial components of the circuit.
+
+        Args:
+            circuit (Circuit): circuit to be trimmed
+            operator (QubitOperator): Operator to trim
+        Returns:
+            trimmed_operator (QubitOperator)
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:137]: No name, just type and description

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            trimmed_circuit (Circuit)
+    """
+    trimmed_circuit, trim_index, trim_states = trim_trivial_circuit(circuit)
+    trimmed_operator = trim_trivial_operator(operator, circuit.width, trim_index, trim_states, reindex=True)
+
+    return  trimmed_operator, trimmed_circuit
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:143]: 1 space after `return`

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
It's a lot of comments, but they're typical of early contributions to a project and simple to fix :) Thank you for opening this !

You have written several functions, and I would like you to write simple tests for each of them to ensure they work (think about corner cases: can they encounter empty circuits or operators ? or tricky situations where the circuit size is enforced by the user, etc, on top of a "normal use case" to check that your thing worked). I came across a few lines in your c ode and I'm honestly not sure about what python do with them, I think that would help.

If you check the "checks" tab of your PR you'll see that we run pycodestyle, which does an amazing job at finding where conventions of style have been broken (extra space or line here and there etc), within a few seconds of you pushing modifications. I caught a few things but definitely rely on this guy next time to get the small stuff out of the way and get the entire test suite run again.

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    trim_index = sorted(trim_index)
+    qu_op_trim = QubitOperator()
+
+    # Calculate expectation values of trivial qubits, update coefficients
+    for op, coeff in qu_op.terms.items():
+        term = pauli_of_to_string(op, n_qubits)
+        c = np.ones(len(trim_index))
+        new_term = term
+        for i, qubit in enumerate(trim_index):
+            if term[qubit] in ('X', 'Y'):
```
[the remark this answers is not in the document]
Comment (AlexandreF-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:46]: `"X" in ("X", "Y")` is indeed returning `True` (the `in` operator works with all collection objects).

[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:46]: changed to 

if term[qubit] in {'X', 'Y'}:

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                c[i] = 0
+            elif term[qubit] == 'Z' and trim_states[i] == 1:
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:48]: changed

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    """
+    trim_states = [x for (y, x) in sorted(zip(trim_index, trim_states), key=lambda pair: pair[0])]
+    trim_index = sorted(trim_index)
+    qu_op_trim = QubitOperator()
+
+    # Calculate expectation values of trivial qubits, update coefficients
+    for op, coeff in qu_op.terms.items():
+        term = pauli_of_to_string(op, n_qubits)
+        c = np.ones(len(trim_index))
+        new_term = term
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:44]: removed new_term

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    circs = circuit.split()
+    e_indices = circuit.get_entangled_indices()
+
+    # Get list of qubits with gates applied to them
+    gated_qubits = [qubit for e_subset in e_indices for qubit in e_subset]
+
+    # Find qubits with no gates applied to them, store qubit index and state |0>
+    zeros = list(sorted(set(range(circuit.width)) - set(gated_qubits)))
+    trim_index = [i for i in zeros]
+    trim_states = [0 for i in range(len(zeros))]
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:80]: changed

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    # Split circuit into components with entangling and nonentangling gates
+    circs = circuit.split()
+    e_indices = circuit.get_entangled_indices()
+
+    # Get list of qubits with gates applied to them
+    gated_qubits = [qubit for e_subset in e_indices for qubit in e_subset]
+
+    # Find qubits with no gates applied to them, store qubit index and state |0>
+    zeros = list(sorted(set(range(circuit.width)) - set(gated_qubits)))
+    trim_index = [i for i in zeros]
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:79]: done!

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+# limitations under the License.
+
+import numpy as np
+
+from tangelo.toolboxes.operators import QubitOperator
+from tangelo.linq import Circuit
+from tangelo.linq.helpers.circuits import pauli_string_to_of, pauli_of_to_string
+
+
+def trim_trivial_operator(qu_op, n_qubits, trim_index, trim_states, reindex=True):
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:22]: done

File: tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+circ = Circuit(mf_gates + pauli_words_gates)
+
+# Reference energy for H4 molecule with single QCC generator
+ref_value = -1.8039875664891176
+
+# Reference indices and states to be removed from system
+ref_trim_index = [0, 2, 4, 6, 8, 9]
+ref_trim_states = [1, 0, 1, 0, 0, 0]
+
+# Reference for which mf_gates are bit fkip gates
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py:59]: typo: flip

File: tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+    def test_is_bitflip_gate(self):
+        """ Test if bitflip gate function correctly identifies bitflip gates """
+        self.assertEqual(ref_bool, [is_bitflip_gate(g) for g in mf_gates])
+
+    def test_trim_trivial_circuit(self):
+        """ Test if circuit trimming returns the correct circuit, states, and indices  """
+
+        trimmed_circuit, trim_index, trim_states = trim_trivial_circuit(circ)
+        self.assertEqual([g for g in ref_circ], trimmed_circuit._gates)
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py:80]: Does `ref_circ._gates` not work as first argument, here?

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from tangelo.toolboxes.operators import QubitOperator
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:17]: group your QubitOperator and count_qubits import on L17

File: tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py (hunk 13 above)
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/tests/test_trim_trivial_qubits.py:80]: it does, thanks!

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        qu_op (QubitOperator): Operator to trim
+        n_qubits (int): number of qubits in full system
+        trim_index (list):  index of qubits to trim
+        trim_states (list): state of the qubits to trim, 0 or 1
+        reindex (bool): If True, remaining qubits will be reindexed
+    Returns:
+        qu_op  (QubitOperator): trimmed QubitOperator with updated coefficients
+    """
+    trim_states = [x for (y, x) in sorted(zip(trim_index, trim_states), key=lambda pair: pair[0])]
+    trim_index = sorted(trim_index)
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:38]: Yeah that probably would make things easier to work with, I'll reformat!
I wasn't sure whether it would be easier for the user to input these as two lists or as a dictionary.

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+            trim_states (list): state of removed qubits, 0 or 1
+            trim_index (list):  index of removed qubits
+
+    """
+    # Split circuit into components with entangling and nonentangling gates
+    circs = circuit.split()
+    e_indices = circuit.get_entangled_indices()
+
+    # Get list of qubits with gates applied to them
+    gated_qubits = [qubit for e_subset in e_indices for qubit in e_subset]
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:103]: that is what i want, thanks!

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        for i, qubit in enumerate(trim_index):
+            if term[qubit] in ('X', 'Y'):
+                c[i] = 0
+            elif term[qubit] == 'Z' and trim_states[i] == 1:
+                c[i] = -1
+
+            new_term = new_term[:qubit - i] + new_term[qubit - i + 1:] if reindex else new_term[:qubit] + 'I' + new_term[qubit + 1:]
+
+        qu_op_trim += np.prod(c) * coeff * QubitOperator(pauli_string_to_of(new_term))
+    return qu_op_trim
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:53]: It is possible, but unlikely! If I do `qu_op_trim.compress()` on the Tangelo QubitOperator then I get a TypeError in the test when using the openfermion function shown below. Do you know a good way to get around that?

```
from openfermion.linalg import qubit_operator_sparse
qubit_operator_sparse(qu_op_trim)
```

 raise TypeError('Operator of invalid type.')
         TypeError: Operator of invalid type.

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    qu_op_trim = QubitOperator()
+
+    # Calculate expectation values of trivial qubits, update coefficients
+    for op, coeff in qu_op.terms.items():
+        term = pauli_of_to_string(op, n_qubits)
+        c = np.ones(len(trim_index))
+        new_term = term
+        for i, qubit in enumerate(trim_index):
+            if term[qubit] in ('X', 'Y'):
+                c[i] = 0
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:46]: Yup you're definitely right, I gave it a shot with break and continue!

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    circuit_new = Circuit()
+    # Go through circuit components, trim if trivial, otherwise append to new circuit
+    for i, circ in enumerate(circs):
+        if circ.width != 1:
+            circuit_new += circ
+            continue
+        # Calculate state of single qubit clifford circuits, ideally this would be done with a clifford simulator
+        # for now only look at first two gate combinations typical of the QMF state in QCC methods
+        num_gates = len(circ._gates)
+        if num_gates not in (1, 2):
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:120]: the` num_gates = 0 ` case is taken care of in L108-109 with  

```
for qubit_idx in set(range(circuit.width)) - set(circuit._qubit_indices):
    trim_states[qubit_idx] = 0
```

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+    # Find qubits with no gates applied to them, store qubit index and state |0>
+    zeros = list(sorted(set(range(circuit.width)) - set(gated_qubits)))
+    trim_index = [i for i in zeros]
+    trim_states = [0 for i in range(len(zeros))]
+
+    circuit_new = Circuit()
+    # Go through circuit components, trim if trivial, otherwise append to new circuit
+    for i, circ in enumerate(circs):
+        if circ.width != 1:
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:114]: added, but kept as "not in (1,2)" since the 0 case is taken care of earlier

`        if circ.width != 1 or circ.size not in (1, 2):
            circuit_new += circ
            continue`

**Review by JamesB-1qbit — APPROVED** [review]

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
We are getting there, I think after those minor comments we can merge

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+
+    if gate.name in {"X", "Y"}:
+        return True
+    elif gate.name in {"RX", "RY"}:
+        try:
+            parameter_float = float(gate.parameter)
+        except (TypeError, ValueError):
+            return False
+
+        return abs(parameter_float - round(parameter_float / np.pi) * np.pi) < 1e-5 and round(parameter_float / np.pi) % 2 == 1
```
[the remark this answers is not in the document]
Comment (elloyd-1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py:85]: The whole thing is checking if it is (or is close to) an odd multiple of pi. I did it this way because I struggled to get the modulo operator to work with a parameter close to pi like 3.14159 and also a multiple of pi. But the following seems to work!

`return abs(parameter_float % (3.14159 * 2) - 3.14159) <= 1e-5`

[the remark this answers is not in the document]
Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/operators/trim_trivial_qubits.py]: Oh yeah, this this would have been my suggestion then (except I would have used np.pi ?). Regarding the "seems" to work: I'd suggest trying different value and making sure that level of precision (1e-5) makes sense to you. If you believe this is working as intended then it's all good and you can resolve this.

I know that hardware has limitations (I recall IonQ can only apply angles up to 1e-3 precision), if you think adding an optional parameter for letting the user set the tolerance is a good idea, I suggest you do it now.

Regardless, please feel free to resolve this when you are satisfied with the behaviour.

File: tangelo/toolboxes/operators/trim_trivial_qubits.py [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    """
+    Check if a gate is a bitflip gate.
+
+    A gate is a bitflip gate if it satisfies one of the following conditions:
+    1. The gate name is either X, Y.
+    2. The gate name is RX or RY, and has a parameter that is an odd multiple of pi.
+
+    Args:
+        gate (Gate): The gate to check.
+        atol (float): The absolute tolerance for gate parameter
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/trim_trivial_qubits.py:70]: Nagging you until the very last change ;)

"Optional, the absolute tolerance for..."

**Review by ValentinS4t1qbit — APPROVED** [review]

**Comment by ValentinS4t1qbit** [comment]
test
